replace package_manager with deb_helper

MarkusTeufelberger opened the pull request:
This is more or less a rewrite from scratch of the existing package_manager rule set (the `dpkg_...` ones). The concepts are very similar (a helper program downloads deb files to a fixed location, renames them and creates a dictionary in a bzl file to make them available to `docker_build` rules at build time), but with several key improvements:

* Every single package is explicitly identified by its path and sha256 hash, making builds reproducible with no implicit external state
* Multiple package sources can be used at the same time, including Ubuntu mirrors or new and old Debian versions. If package names collide, there are no issues as long as they are not from the same source.
* It is possible to refer to several versions of the same package
* A helper script is included that can run the equivalent of dist-upgrade automatically on all packages that are defined
* The helper script can also parse all BUILD files for docker_build rules that contain references to debs and automatically add these to the WORKSPACE file as well
* The helper script can also be used to pin versions (as well as always updating to `latest`)

## Issues:
The deb_loader executable currently is loaded from localhost (deb_helper/deb_helper.bzl line 7) since the binary obviously is not hosted by Google (yet). This needs to be changed to something similar as in the package_manager rule (`https://storage.googleapis.com/distroless/package_manager_tools/` etc.) before merging this. A cloudbuild.yaml file is already provided.
I haven't found a way how to make Bazel build parts of its toolchain before actually starting a build, so it might be necessary (for now) to host the binary somewhere, similar to what package_manager did.

The helper script does not yet check the gpg signature(s) of the `Release` file. This will be fixed in the future, currently package_manager doesn't check signatures either anyways and it trusts whatever snapshots.debian.org provides. It is a bit difficult since it should also be possible to verify signatures of internal Debian mirrors that are not signed by people in the Debian/Ubuntu projects.

Debian mirrors move forward! Over time you can expect that some packages are no longer available on all mirrors because some new versions are released. Solutions are to either use the snapshot mirror (the currently used solution), update to current versions of packages with the helper script or to maintain a private mirror that hosts the files you use in your builds. This is out of scope for this rule set + helper script, though maybe in the future a script that parses WORKSPACE files and automatically pushes the referred files to a private mirror might be possible.

## References:
See the discussions in #72 and #68 about why the WORKSPACE file looks a little bit more ugly yet it still improves the project imho.

Looking forward to your feedback! :-)

r2d4 commented:
I'm definitely open to changing any of the implementation details around these rules, but I fail to see how this is an improvement, save for the helper script.  All the points you listed are currently possible with the current rules

> Every single package is explicitly identified by its path and sha256 hash, making builds reproducible with no implicit external state

The current rules not only check the checksum of the file, but also the checksum of the package list.  This makes things reproducible.  

> Multiple package sources can be used at the same time, including Ubuntu mirrors or new and old Debian versions. If package names collide, there are no issues as long as they are not from the same source.

This is already true by using multiple `dpkg_src` rules.

> It is possible to refer to several versions of the same package

You can already do this.  I think that we should definitely add something more explicit here, but by using a particular `dpkg_list` rule with a particular `dpkg_src` rule, you are inherently picking a specific version.  You can use as many of these rules together as you wish.

> A helper script is included that can run the equivalent of dist-upgrade automatically on all packages that are defined

I actually like the idea of a helper script here to generate some of the rules automatically, this might be particularly useful for searching for snapshots with particular versions.

MarkusTeufelberger commented:
> The current rules not only check the checksum of the file, but also the checksum of the package list. This makes things reproducible.

They depend on snapshot.debian.org being available and serving the correct Package.gz file (which contains the hashes) at build time. It also will have to serve all packages.
It is currently not the case that hashes of individual packages are easily available, but I agree that if you look at a single slice in time, your build will be reproducible (_if_ the snapshot service is available).

> This is already true by using multiple dpkg_src rules.

And a different dpkg_list rule for each source, yes. I don't see how Ubuntu packages or private mirrors of internal packages are possible though?

> You can already do this. I think that we should definitely add something more explicit here, but by using a particular dpkg_list rule with a particular dpkg_src rule, you are inherently picking a specific version. You can use as many of these rules together as you wish.

Debian mirrors (e.g. using aptly, here's an example for you to test: http://repo.aptly.info/dists/squeeze/main/binary-amd64/Packages) can and do host more than one version of the same package at the same time. This is also a requirement for our containers. How does listing a different source mirror help there?

MarkusTeufelberger commented:
Oh, and the helper program as well as the rules are by design named differently. Another option would be to just add them as alternative to the project. The pr here is also more to showcase the concept and tools, I'm more than willing to rework it and have a discussion.

mattmoor commented on java/BUILD:
```diff
@@ -1,14 +1,15 @@
-        packages["zlib1g"],
-        packages["openjdk-8-jre-headless"],
+        debian_jessie_amd64["zlib1g"],
+        debian_jessie_backports_amd64["openjdk-8-jre-headless"],
```
Is it possible today to split some of this renaming into its own change?

MarkusTeufelberger replied:
Yes, definitely.

mattmoor commented on deb_helper/deb_packages.bzl:
```diff
@@ -0,0 +1,40 @@
+def _deb_packages_impl(repository_ctx):
```
This file could use a license header.

MarkusTeufelberger replied:
Yours doesn't have one either, so I didn't add one. Can be done of course.

mattmoor commented:
@MarkusTeufelberger thanks for the PR, I very much appreciate the thought and effort you have put into this, and the discussions you are prompting.

I am still not 100% sure how I feel about this approach vs. what we have (I think some of this is a more fundamental question/debate of what role `WORKSPACE` plays in the Bazel ecosystem: dependency fetching vs. dependency resolution).  @damienmg @ulfjack FYI.

Looking at this independent of that debate, I left a few comments / suggestions.  I have not read through everything :)

r2d4 commented:
I would love the support private and other "rolling" public apt repositories besides the snapshot archive, but until we get gpg signing, I don't think its possible.  

The current rules don't check the gpg signature either, but they do verify the checksum of the packages archive file, which is equivalent (maybe even a narrower form of trust, since you are only trusting one particular revision of a particular snapshot).

If you don't check the checksum or the signature of the package archive, the checksums are practically meaningless (since they are autogenerated and self reported from an insecure source).  I think its especially bad since we are pulling the package archive over HTTP and not even HTTPS.

r2d4 commented:
I think another behavior we need to discuss for is that this change would require indefinite maintenance.  Using any sort of rolling release will eventually break the build - (not if, but when) packages get updated.  

You could simply run the update helper, but someone is going to need to do that, send the PR, approve, etc before any new contributions are made.

mattmoor commented:
@r2d4 I think that this is sort of WAI for reproducible builds.

We could certainly automate something to run ~daily to update the rules and open a PR?

r2d4 commented:
Who is going to review 365 PRs a year?  Pushing automatically is a little scary.  That being said, I'm really against having an open source repository that needs continuous maintenance to be build-able.  AFAIK there are no examples of this out in the wild, probably for good reason.

To me, reproducible builds not only includes repeated builds giving the same results, but also being able to go back in time and get the same build artifacts.  I could imagine a user trying to reproduce some previously released artifacts, and being unable to.  Sure, with the checksums, you might be able to find and provide the right packages.  But building at a certain point in time is no longer as easy as `git checkout && bazel build ...`.

mattmoor commented:
I wouldn't expect daily PRs (maybe once a week, is there any way we can gather data?), and I obviously meant pushing to a fork (the bot shouldn't have write access).

I think you are conflating hermetic builds with reproducible builds.  We solve this internally with hermeticity, but frankly with external dependencies (even pinned, which I'd call pseudo-hermetic) you run the risk of this kind of breakage unless you float dependencies forward (not pseudo-hermetic, not reproducible).

External folks generally cope with this by either floating dependencies forward (bad for reproducibility), vendoring (bad for binary deps), mirroring / snapshotting (potentially a heavy process), or building from [vendored] source (complex, a burden for build time, presents its own upkeep challenges).

I think having the roll forwards be intentional and controlled is a good thing, and I think we can make it nearly push-button (e.g. automatically generate a PR).  As it stands, we would only regenerate images when a new commit is made, so floating the source deps forwards is insufficient to patch security vulnerabilities.

You are right, that folks often opt to not require continuous maintenance just to build, but that doesn't mean continuous maintenance isn't necessary.  One of the biggest problems with `vendor`ing and `third_party` is that they tend to lag behind the upstream, which can be a serious security problem.


In the context of this PR, it seems like the approach that might appeal to you the most would be to make the workspace-update actually download the debs and mirror them to a public GCS bucket (our own snapshot), and then simply use `new_http_file` with a checksum option to download them (likely still wrapped so we can rewrite things).  However, that makes updating the package list a privileged operation, which is its own wart.

r2d4 commented:
I still think required automation, on any interval, to keep the repository usable is a no-go.  That multiples the complexity.  `interval of updates * number of repositories that use these rules` PRs have to be reviewed and pushed.  Now those repositories also need to implement automation to keep their projects build-able.

> You are right, that folks often opt to not require continuous maintenance just to build, but that doesn't mean continuous maintenance isn't necessary. One of the biggest problems with vendoring and third_party is that they tend to lag behind the upstream, which can be a serious security problem.

We should definitely provide the scaffolding needed to update packages and patch CVEs.  However, I don't think forcing users to accept rolling updates makes sense.  This would eventually break users who need to pin to a certain version.  

Hermetic or reproducible, mirroring is equivalent to a snapshot method (we don't care who hosts it, as longs as its the same artifact).

mattmoor commented:
I think you are assuming folks are taking a source dependency, which the
way the repo is setup is a nightmare (all of the stuff we have would need
to be redeclared in their own workspace).

Folks can trivially pin at the image level and we have no immediate plans
to start retiring those images.  So afaik we are taking this on ourselves
and users can float by taking a tag dep, or pin by taking a digest dep.

Are you aware of any source level dependencies on this repo?

On Wed, Aug 30, 2017, 11:35 AM Matt Rickard <notifications@github.com>
wrote:

> I still think required automation, on any interval, to keep the repository
> usable is a no-go. That multiples the complexity. interval of updates *
> number of repositories that use these rules PRs have to be reviewed and
> pushed. Now those repositories also need to implement automation to keep
> their projects build-able.
>
> You are right, that folks often opt to not require continuous maintenance
> just to build, but that doesn't mean continuous maintenance isn't
> necessary. One of the biggest problems with vendoring and third_party is
> that they tend to lag behind the upstream, which can be a serious security
> problem.
>
> We should definitely provide the scaffolding needed to update packages and
> patch CVEs. However, I don't think forcing users to accept rolling updates
> makes sense. This would eventually break users who need to pin to a certain
> version.
>
> Hermetic or reproducible, mirroring is equivalent to a snapshot method (we
> don't care who hosts it, as longs as its the same artifact).
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/distroless/pull/94#issuecomment-326080526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACVE4t8TnDrGJuhkXikRyU91cASLhJcrks5sdauAgaJpZM4PGO2f>
> .
>

r2d4 commented:
A quick search shows istio/mixer using the package_manager rules https://github.com/istio/mixer/blob/master/WORKSPACE#L554.  

I think what you say is true for the docker_rules and the distroless pattern, but I expect everyone to import the package_manager rules at the source level.  The package_manager rules have much more value outside creating the base images.  In fact, they have little value for the base images (where they rarely change), but more value in creating minimal images that consume packages.

If these rules are for our exclusive use as image publishers, this proposed change makes no sense (why support arbitrary and private apt repositories if we will never bundle any of those .debs in our base images?)

MarkusTeufelberger commented on deb_helper/README.md:
````diff
@@ -0,0 +1,192 @@
+
+```python
+load(
+    "//deb_helper:deb_helper.bzl",
````
Sure, what would you prefer as a name instead?

It evolved a bit from starting out as only the helper tool to rewrite the WORKSPACE file, then also needing to rewrite the package_manager portion and actually removing the rewriting tool from the build process, hence the slightly undescriptive name.

MarkusTeufelberger commented on deb_helper/README.md:
```diff
@@ -0,0 +1,192 @@
+    "deb_loader",
+    "deb_packages",
+)
+deb_loader()
```
So `deb_loader()` should be `deb_repositories()`?

MarkusTeufelberger replied:
I settled on

```
load(
"//deb_loader:deb_loader.bzl",
"deb_repositories",
"deb_packages"
)
```

Or would you prefer `deb_loader_repositories`?

MarkusTeufelberger commented on deb_helper/README.md:
````diff
@@ -0,0 +1,192 @@
+Hashes are checked after downloading files from a mirror.
+
+```python
+deb_packages(
````
It is only described in the tool section of the readme, so I agree that the manual process (look up the values in a `Packages` file that you have manually verified to be trustworthy) should be documented too here.

MarkusTeufelberger commented on deb_helper/README.md:
```diff
@@ -0,0 +1,192 @@
+        "python2.7-minimal": "pool/main/p/python2.7/python2.7-minimal_2.7.9-2+deb8u1_amd64.deb",
+        "zlib1g": "pool/main/z/zlib/zlib1g_1.2.8.dfsg-2+b1_amd64.deb",
+    },
+    packages_sha256 = {
```
> Bazel should support [...]

I agree, currently it seems however that only the following attributes are supported: https://docs.bazel.build/versions/master/skylark/lib/attr.html

Initially I wanted to have a more complex object here, like this:
```
"emacs" = {
  "distro" = "jessie",
  "path" = "pool/foo/bar",
  "sha256" = "12345678abcd"
  ...
}
```
which definitely doesn't seem to be possible at all. I couldn't find ANY instance of string_list_dict being used in the wild (string_dict is already quite exotic and not even supported by `buildozer` for example) and having a list where the individual items HAVE to be in a certain order ("distro" first, then "sha256" or whatever) also didn't seem appealing.

Hence the decision to split the single rule into one rule per source and to split the dictionary in two.

I'll look into `Fileset`, where is it documented?

MarkusTeufelberger commented on deb_helper/README.md:
```diff
@@ -0,0 +1,192 @@
+`update_workspace` is exactly such a helper program.
+It uses the `buildifier` and `buildozer` tools from [https://github.com/bazelbuild/buildtools](https://github.com/bazelbuild/buildtools), which need to be available on your $PATH.
+
+To update hashes and paths in a single `deb_packages` rule with the `update_workspace --update/-u RULENAME` switch.
```
True, seems reasonable and better fitting to make everything available to a `bazel` invocation.

MarkusTeufelberger replied:
Looked deeper into it and I think it might be reasonable to just always run the `--add` functionality when invoked that way which adds missing packages and updates the paths/hashes. It also could remove packages that are not referred in docker_build rules, but I guess there might be other use cases for having a *.deb file available in your build environment, such as the cacerts example.

MarkusTeufelberger replied:
Revisiting this one... it is surprisingly hard to package Python executables that depend on external packages. Python-apt does not make the job easier at all (they don't publish to pypi, contain C bindings and of course are not using Bazel to build).

I'll probably need to write a BUILD file for the project there and then also stop using subpar who don't support C dependencies yet. In theory the approach from gazelle works (and I have it running the binary already, it'll just fail as soon as it is supposed to parse a file), but to get it done with Python is tougher than expected.

MarkusTeufelberger commented on WORKSPACE:
```diff
@@ -10,59 +10,94 @@ load("@io_bazel_rules_go//go:def.bzl", "go_repositories")
-    arch = "amd64",
+# For every source of deb packages create a deb_packages rule
+# The name does not have to be like the one here, but it helps to know which source is used in BUILD files
+deb_packages(
```
Agreed.

MarkusTeufelberger commented:
Thanks for all the comments and discussion so far! :+1: 

To go point by point:

> The current rules don't check the gpg signature either, but they do verify the checksum of the packages archive file, which is equivalent (maybe even a narrower form of trust, since you are only trusting one particular revision of a particular snapshot).

True, I really hope that dlorenc (who did the commit in a browser) did check the file with gpg on his computer before submitting https://github.com/GoogleCloudPlatform/distroless/commit/bd16e2028cc0dd6acba3de58448c94b3d2ead21a and that mattmoor did the same before approving the PR.

As I wrote, gpg verification is coming. Until then, you can also manually verify - it is only relevant for the helper tool, not the actual WORKSPACE rules anyways.

> Using any sort of rolling release will eventually break the build - (not if, but when) packages get updated. 

True, to prevent the scenario you are suspecting for this repository, snapshot.debian.org could be used as package source with a second CI job that uses up-to-date mirrors instead. This is how I would recommend this project to be used anyways: Have an internal Debian mirror that offers multiple package versions including older ones and 2 builds - one stable for production and one canary build for latest packages. Whenever necessary, update the stable build to newer versions that you already know are working.

As I wrote above: "Solutions are to either use the snapshot mirror (the currently used solution), update to current versions of packages with the helper script or to maintain a private mirror that hosts the files you use in your builds." - currently you are hard coding solution 1, entering a certain state of snapshot.debian.org as mirror in the ruleset here would have the same effect but allow for solutions 2 and 3 too.

The current behaviour outsources all of this functionality to a third party project with no SLA that I could find, the latest news item from 3 years ago and some minor bugs (as can be expected with such a service: https://bugs.debian.org/cgi-bin/pkgreport.cgi?pkg=snapshot.debian.org;dist=unstable). Once they go down, nothing can be built any more, even if you have all packages available locally, since you can't verify their hashes any more without the Packages file from the snapshot server.

> Are you aware of any source level dependencies on this repo?

Busybox, Jetty and Node are downloaded from source more or less, right? At least they have to be manually updated, just like bazel rule sets.

> If these rules are for our exclusive use as image publishers, this proposed change makes no sense

I really hope this project is not just "for Google internal use only". I anyways also would be more than happy to have the functionality in this PR next to the current one instead of replacing it. Since it is a superset of functionality (other than the ease of editing WORKSPACE by hand) and I wanted to have more focused discussions, I proposed to replace the existing stuff. If you would be more comfortable with just upstreaming the rules and want to depend on snapshot.debian.org at build time for now, this is also fine by me. I would just be not too thrilled if I need to maintain a fork of this project, because I believe it is useful beyond Google internal use.

MarkusTeufelberger commented:
I'll work on addressing the feedback by @mattmoor and will push over the current commit to split the changes into several commits. If you want to, I can also create a second version that merges into the current repo instead of replacing the current rule set, so you can try it out side by side.

r2d4 commented:
> As I wrote, gpg verification is coming. Until then, you can also manually verify - it is only relevant for the helper tool, not the actual WORKSPACE rules anyways.

I think the point is that you can't verify.  If you don't check the signature OR the checksum of the package.gz file, the checksums it reports are meaningless (which are then frozen to your workspace with the helper tool).  You need to do one, or the other.  The current rules are limited to the checksum, but I think we both agree the end goal is to use signature verification.

I'm going to distill this discussion into a few points I think are reasonable

* this repo should always use the snapshot package. 
* signature verification should be the end goal 
* It would be nice to be able to use arbitrary sources other than the snapshot source
* There needs to be a method of pinning packages other than just name + source.

**PR Feedback:**
My overall feedback is that a complete rewrite of all these rules should be the last resort.  In fact, I think that this PR could be implemented in very few lines.

* Besides the helper script and some URL parsing, the only real change here is the update script and the fact that the bazel package list array now contains a checksum which is used to verify the contents rather than from the package list itself.
* I think that GPG verification is an absolute requirement before we support arbitrary package sources.

Other feedback:
* Keep pointing at the snapshot, I don't think we want to deal with the machinery needed to auto update this repository.
* Remove all superfluous renames.  I think this would give us a better opportunity to see the actual changes.  The `deb_helper/deb_file_loader.py` rule does almost exactly the same thing as the current `dpkg_parser.py`, except for sha verification of the package list.  `architecture` instead of `arch` may be your style -  the files are definitely candidates for being reworked -  but changes like this make it very hard to review the merits of this PR.  Open up other PRs for renames and field changes.

dlorenc commented:
Echoing @r2d4 and @mattmoor, thanks a lot for starting this discussion! I was never terribly happy with the snapshot mirror requirement, agree with many of the problems you raised about it.

It was really just the only thing we could come up with to give a good combination of reproducibility/security at the time.

One other option I can think of is to support more robust vendoring. What if we made it easier to fetch, commit and use Packages.gz files in a repo, as well as the resulting .debs? That would help mitigate some of the issues with relying on the snapshot service being available.

MarkusTeufelberger commented:
> You can't verify.

You can easily manually verify a `Packages.gz` file with gpg (just as dlorenc and mattmoor hopefully already did when they approved https://github.com/GoogleCloudPlatform/distroless/pull/87 and also what you yourself did in https://github.com/GoogleCloudPlatform/distroless/commit/2c24ffbf04b1fab7f38a9c31497f20d9348641e9 when you came up with the hash `8ff5e7a54d4e75bbbcd2f43ebc7cb4a082fbc5493bc9fb2dcdaaeacba6e76dee` - you did verify the Release file actually is signed by the keys on https://ftp-master.debian.org/keys.html, right?), then extract its contents, look up the hashes/patches manually and add them to the WORKSPACE file. The update_workspace tool just automates these steps, since as soon as you have the tuple `(package_name, version, architecture)`, you can determine if and where a package is available on a Debian mirror.

> this repo should always use the snapshot package.

I would argue that Google should host the deb files they put in their containers themselves and only fallback on the snapshot mirror, but yes, you could make sure that historic commits build if you depend on the snapshot mirror, if that is a priority for this project.

> signature verification should be the end goal

For tools that process `Packages.gz` files automatically, yes.

> It would be nice to be able to use arbitrary sources other than the snapshot source

I would say that it is critical, but yes, it would be nice too. :-)

> There needs to be a method of pinning packages other than just name + source.

:+1:
`(package_name, version, architecture)` is the unique identifier for a package, and a checksum must be provided too to make sure the mirror is not malicious. Having `(distribution_name distribution_type)` in there as well makes is easier to also support Ubuntu packages for example.

> In fact, I think that this PR could be implemented in very few lines.

The current rule set only hashes the Packages.gz file which will change frequently on non-snapshot mirrors, even if all files referenced in `dpkg_list` rules are still the same. This means without some major rearchitecture they are no longer suitable for actual mirrors like deb.debian.org (which was used before).

Another issue I have with the current rules is that they hide a lot of important information (file hashes + package version) in the Packages.json file. This makes auditing which container contains which exact binary a nightmare. You need to manually create and parse a Packages.json file to understand what's going on - after all the containers themselves intentionally don't have any tooling inside to help with this in any way.

Lastly the updater tool also has a few lines of code.

> Keep pointing at the snapshot, I don't think we want to deal with the machinery needed to auto update this repository.

Will do, noted for the refactor.

> Remove all superfluous renames. [...] Open up other PRs for renames and field changes.

I intentionally tried to not build upon `dpkg_parser.py` to make sure that I can maintain a clean fork, if necessary. Also I don't need most of its functionality. If you are willing to accept modifications in preparation to accepting this PR, I'm more than willing to submit some PRs to add the functionalities I need there (mirror support, path and hash as parameters, support for ubuntu, rename debs to sha256 instead of base64 to ensure higher uniqueness) instead. I would also use probably use http://apt.alioth.debian.org/python-apt-doc/library/apt_pkg.html to parse the `Package` file (`parse_package_metadata` can't handle some valid `Packages` files), but in my use case this is not done at build time anyways.

> It was really just the only thing we could come up with to give a good combination of reproducibility/security at the time.

I agree, it was a quick win looking at the architecture of the `dpkg_` rules compared to referring to debs individually as `http_file`s like before. I set out in the beginning to just (re)write the http_file rules, but I really liked the more descriptive style of the more dedicated rules. I am just afraid that the design of the current rules would solidify into something that doesn't even know which version a package will have before a build is actually started without querying an external service, hence the rather fast and rather large contribution that unfortunately seems to overthrow some of the existing stuff.

> One other option I can think of is to support more robust vendoring. What if we made it easier to fetch, commit and use Packages.gz files in a repo, as well as the resulting .debs? That would help mitigate some of the issues with relying on the snapshot service being available.

These repositories are called "mirror". ;-)
If this gets merged, the next tool I probably would write is a tool to fetch and push all referred deb files to a mirror that you can host yourself (this is extremely easy with lots of good tools out there), so you don't have to rely on external infrastructure. Downloading and parsing a whole Packages file (they are not exactly small...) at build time still seems like a design error to me though, even it it is done from a local or highly available mirror.

---

Anyways, for now I'll rewrite some parts of the tooling (e.g. to be able to use the updater more like `gazelle`) and it is not unlikely that there will be another field added to the `deb_packages` rule that refers to gpg keys. I'll also split the commit up into more manageable logical chunks (probably along the lines of "add new rule", "add updater tool", "add sample container built by new rule" with the option of removing the existing rules + rewriting BUILD files). Tomorrow I'll probably have to work on a different project, so please bear with me over the weekend for these changes to land and sorry for dumping this on you a bit unannounced and in a slightly messy fashion.

MarkusTeufelberger commented on WORKSPACE:
```diff
@@ -10,59 +10,94 @@ load("@io_bazel_rules_go//go:def.bzl", "go_repositories")
-    url = "http://snapshot.debian.org/archive",
+    distro_type = "debian",
+    mirrors = [
+        "http://deb.debian.org",
```
Typically mirrors are specified up to the root folder in `sources.list` files (e.g. `deb http://deb.debian.org/debian unstable main`). For now I went with the simpler way this was done in this project before, but it might be the case that this has to change in the future.

MarkusTeufelberger commented:
Closing this PR in favor of the more incremental PR #96, sorry for dumping this big change at once onto you.
